{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9772f56c-3ef8-4e5a-894c-5810e82f0903",
   "metadata": {},
   "source": [
    "## About\n",
    "This notebook does a few things for preparing for proteome-wide analysis.\n",
    "\n",
    "1. Generates the file `ysn_only.txt` which can be mapped to uniprotIDs on using gene name as the input (and restricting to S. cerevisiae).\n",
    "2. Generates the file `ysn2uniprot\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "27c3fb43-bc0a-49f3-86cd-979c72b94f35",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbf5ee40-e355-4574-beb7-1daecac1bc31",
   "metadata": {},
   "source": [
    "### Build the YSN list \n",
    "This first cell builds a text file with all the S. cerevisiae Yeast Systematic Names (YSNs) for the yeast genes. Note we include this for completeness BUT there is no need to re-run this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "700ae299-656e-4eb1-871b-14721d42de42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in all proteomes....\n",
      "... DONE!\n",
      "Reading in 5436 separate FASTA files - this may take 30-40 seconds...\n",
      "... DONE!\n"
     ]
    }
   ],
   "source": [
    "USE_HHL_TOOLS = True\n",
    "\n",
    "if USE_HHL_TOOLS:\n",
    "\n",
    "    WRITE_OUT_YSN = True\n",
    "\n",
    "    from yeastevo import Pillars\n",
    "\n",
    "\n",
    "    # build and read in the sequence information\n",
    "    fungi_matrix = Pillars()\n",
    "\n",
    "    # get the set of IDs and read in alligned sequences\n",
    "    all_valid_IDs = fungi_matrix.all_aligned_scerevisiae_YSNs()\n",
    "\n",
    "    if WRITE_OUT_YSN:\n",
    "        with open('generated_data/ysn_only.txt','w') as fh:\n",
    "            for k in all_valid_IDs:\n",
    "                fh.write(f\"{k}\\n\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cfa14a9-3dc8-4047-804a-9457575e3d16",
   "metadata": {},
   "source": [
    "## Build YSN to UniProt mapping\n",
    "The code here generates a comprehensive mapping of each of the YSN identifiers to:\n",
    "\n",
    "1. UniProt ID\n",
    "2. Protein name\n",
    "3. List of Gene names (many yeast genes have multiple names).\n",
    "\n",
    "This cell depends on the `ysn_only.txt`file indirectly, in that that file must first be generated and then is passed to the UniProt ID mapping service to generate the `idmapping_2024_08_28.tsv` file, which is what the cell below is parsing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7363cc11-8a1e-4add-8011-b8108a5faa36",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('input_data/idmapping_2024_08_28.tsv', 'r') as fh:\n",
    "    raw_uniprot_mapping = fh.readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "81e7c6ce-ecc7-43f8-b67c-2e021ceb5d30",
   "metadata": {},
   "outputs": [],
   "source": [
    "ysn2info = {}\n",
    "for line in raw_uniprot_mapping[1:]:\n",
    "    sline = line.strip().split('\\t')\n",
    "\n",
    "    ysn = sline[0]\n",
    "    uid = sline[1]\n",
    "    name = sline[2]\n",
    "    gene_names = sline[3].split()\n",
    "\n",
    "    if ysn in ysn2info:\n",
    "        raise('Error - found duplicate')\n",
    "    ysn2info[ysn] = [uid, name, gene_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3a5a5a6e-27e5-4835-acb3-eb392ed9bf7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate PKL file for easy computational use\n",
    "with open(f'generated_data/ysn2uniprot.pkl', 'wb') as file:    \n",
    "    pickle.dump(ysn2info, file)\n",
    "\n",
    "# generate text version for easy reading\n",
    "with open(f'generated_data/ysn2uniprot.tsv', 'w') as file:\n",
    "    file.write('YSN\\t UniProtID\\t Name\\t Gene Names\\n')\n",
    "    for ysn in ysn2info:\n",
    "        file.write(f\"{ysn}\\t{ysn2info[ysn][0]}\\t{ysn2info[ysn][1]}\\t{\", \".join(ysn2info[ysn][2])}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
