{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1de1d575-70f0-4a41-b3be-3a19f000dca6",
   "metadata": {},
   "source": [
    "## About\n",
    "This notebook uses the precomputed full protein sequence alignments to extract out parlogous/orthologous IDRs. This notebook unavoidably make extenisve use of internal holehouse lab tools, to the extent that extracting out that code from this notebook is not trivial.\n",
    "\n",
    "IF you are reading this and you want intermediate data that you can't access, please message Alex and he'll re-write this in a way that intermediate data is read in instead of being calculated. However, this would require a major re-restructuring of the code so I didn't want to spend the time doing this until someone needed it. Happy to do it though :)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4156893-fe76-4670-a243-f86a059a0e3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "from tqdm import tqdm\n",
    "import pickle\n",
    "\n",
    "import protfasta\n",
    "import metapredict as meta\n",
    "\n",
    "from finches import Mpipi_frontend\n",
    "from finches.data import fingerprints\n",
    "mf = Mpipi_frontend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59419e86-f2e3-4818-a17e-21514b6d2ce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from yeastevo import Pillars\n",
    "from housetools.sequence_tools.sequence_conservation import ConservationCalculator\n",
    "from housetools.interfaces.eggnog import nogtools\n",
    "import yeastevo\n",
    "\n",
    "# activate and pre-load the conservation object\n",
    "CC = ConservationCalculator()\n",
    "\n",
    "# build and read in the sequence information\n",
    "fungi_matrix = Pillars()\n",
    "\n",
    "# get the set of IDs and read in alligned sequences\n",
    "all_valid_IDs = fungi_matrix.all_aligned_scerevisiae_YSNs()\n",
    "\n",
    "# build mapping of YSN to amino acid sequence with for. cerevisiae \n",
    "ysn2seq = {}\n",
    "for ysn in all_valid_IDs:    \n",
    "    full_seq = fungi_matrix.get_orthologs(ysn, silent=True)['Scerevisiae'][0][1]    \n",
    "    ysn2seq[ysn] = full_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc608c8a-e727-400e-a3ae-53ceebd38f61",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OPTIONS - options that can change script output defined here\n",
    "metapredict_version = 'v3'\n",
    "\n",
    "# shortest IDR (sans alignment chars) for \n",
    "min_idr_length = 50\n",
    "\n",
    "# minimum level of disorder we'll accept for an orthologous \"IDR\" (identified by excising equvalent positions based on S. cereviase IDR)\n",
    "fraction_disordered = 0.8\n",
    "\n",
    "# defines is we generate output pickle files with globally aligned or unaligned sequences\n",
    "remove_gaps = True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "942f6a1a-e9a7-43ba-aebc-2c946fa9475c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# predict ysn to DisorderObject objects so we can map IDR boundaries - note this will take ~20-30 seconds for metapredict\n",
    "ysn2disorder = meta.predict_disorder(ysn2seq, \n",
    "                                     show_progress_bar=True, \n",
    "                                     return_domains=True, \n",
    "                                     version=metapredict_version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f230c72a-1be0-4d44-84b1-371d34732f0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_orthologous_idrs(ysn, fungi_matrix, ysn2disorder, min_idr_length=50, fraction_disordered_threshold=0.8):\n",
    "    \"\"\"\n",
    "    Extract out orthologous IDRs using S. cerevisiae as the reference frame. Algorithmically, this\n",
    "    function does the following:\n",
    "\n",
    "    1. Using S. cerevisiae as the reference, extracts the start/end of each IDR in the protein\n",
    "    associated with the input gene name (ysn).\n",
    "\n",
    "    2. Gets the corresponding subregions from the orthologs. Specifically this translates the real-sequence\n",
    "    indexing from S. cerevisae into the aligned-sequence positions, and then extracts out all the positions\n",
    "    from the orthologs in the aligned space, and then removes the alignment character.\n",
    "\n",
    "    3. Filter those orthologous IDRs. First for each orthologous IDR, we remove sequences where there is no IDR.\n",
    "    Second, we predict disorder for those orthologous IDRs, and then only keep IDRs that are both >80% disordered\n",
    "    and over 50 amino acids. The 80% and 50 amino acids can be changed by the fraction_disordered and min_idr_length\n",
    "    parameters\n",
    "\n",
    "    Finally, we return a dictionary where the keys are are the yeast gene names, and the values are also dictionaries.\n",
    "    For the value dicts, the keys are IDR_X_START_END where X is an identifier that starts at 1 and is incremented, and\n",
    "    START and END are the positions in the original S. cerevisae protein where the IDR started. The values are ANOTHER\n",
    "    dictionary, here here, kere are gene_name_species_gene_name and values are the sequence. So in short the return\n",
    "    structure is\n",
    "\n",
    "    \n",
    "    idr_dict_filtered[IDR_IDENTIFIER][SPECIES_IDENTIFIER] = IDR from species\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "\n",
    "    ysn : str\n",
    "        Name of the yeast \n",
    "\n",
    "    fungi_matrix : yeastevo.pillars.Pillars\n",
    "        Pre-constructed and initiallized Pillars object from the yeastevo repository\n",
    "\n",
    "    ysn2disorder : dict\n",
    "        Dictoinary that maps the ysn to a DisorderObject, which holds the disorder \n",
    "        domains and boundaries\n",
    "\n",
    "    min_idr_length : int\n",
    "        Shortest IDR length we'll use\n",
    "\n",
    "    fraction_disordered_threshold :float\n",
    "        Minimum fraction disorder we will tolerate for an IDR of interest.\n",
    "\n",
    "\n",
    "    Returns\n",
    "    -----------\n",
    "    dict\n",
    "        Returns a dictionary where keys are are the yeast gene names, and the values \n",
    "        are also dictionaries. For the value dicts, the keys are IDR_X_START_END where \n",
    "        X is an identifier that starts at 1 and is incremented, and START and END are \n",
    "        the positions in the original S. cerevisae protein where the IDR started. \n",
    "        The values are ANOTHER dictionary, here here, kere are gene_name_species_gene_name \n",
    "        and values are the sequence. \n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    if fraction_disordered_threshold < 0 or fraction_disordered_threshold > 1:\n",
    "        raise Exception('Fraction disordered only makes sense between 0 and 1')\n",
    "\n",
    "    if min_idr_length < 1:\n",
    "        raise Exception('Shortest IDR length only makes sense as a positive integer')\n",
    "\n",
    "    # get the orthologous IDRs\n",
    "    seq_dict = fungi_matrix.get_aligned_scerevisiae_sequences(ysn)\n",
    "\n",
    "    ## get the S. cerevisae protein key from the seq_dict\n",
    "    # Note this warns if we find multiple Scerevisiae keys but always choose the \n",
    "    # first in the list.\n",
    "    possible_yeast_indices = [x for x in seq_dict.keys() if x.find('Scerevisiae') >-1]    \n",
    "    \n",
    "    if len(possible_yeast_indices) == 1:\n",
    "        reference_key = possible_yeast_indices[0]\n",
    "    elif len(possible_yeast_indices) > 1:        \n",
    "        reference_key = possible_yeast_indices[0]\n",
    "        print(possible_yeast_indices)\n",
    "        print(f'Found multiple S. cerevisiae indices; using {reference_key}')\n",
    "    else:\n",
    "        raise Exception('Should not get here...')\n",
    "\n",
    "    # buld out the ortholgous IDRs\n",
    "    idr_idx = 1\n",
    "    idr_dict = {}\n",
    "    idr_dict_aligned = {}\n",
    "\n",
    "    # for each IDR in S. cerevisiae\n",
    "    for boundaries in ysn2disorder[ysn].disordered_domain_boundaries:\n",
    "\n",
    "        # get the reference key for the yeast protein        \n",
    "        start = boundaries[0]\n",
    "        end  = boundaries[1]-1 # note we do a -1 because the boundaires are in Python slice syntax\n",
    "\n",
    "        # extract out the orthologous IDRs that \n",
    "        idr_dict[f\"IDR{idr_idx}_{start}_{end}\"] = nogtools.get_alignment_subregion_v2(seq_dict, \n",
    "                                                                                        reference_key, \n",
    "                                                                                        start=start, \n",
    "                                                                                        end=end)\n",
    "        \n",
    "\n",
    "        idr_dict_aligned[f\"IDR{idr_idx}_{start}_{end}\"] = nogtools.get_alignment_subregion_v2(seq_dict, \n",
    "                                                                                              reference_key, \n",
    "                                                                                              start=start,\n",
    "                                                                                              end=end,\n",
    "                                                                                              remove_gaps=False)\n",
    "                                                                                                  \n",
    "        \n",
    "            \n",
    "        \n",
    "        idr_idx = idr_idx + 1\n",
    "\n",
    "    # now we filter through, note we ALWAYS include S. cerevisiae\n",
    "    idr_dict_filtered = {}\n",
    "\n",
    "    \n",
    "    # for each IDR in our newly built IDR dictionary\n",
    "    for d in idr_dict:\n",
    "\n",
    "        # remove orthologs where no IDRs were found\n",
    "        tmp = {}\n",
    "        tmp_aligned = {}\n",
    "        for k in idr_dict[d]:\n",
    "\n",
    "            # WEIRD situation where for some reason the cerevisiae IDR is length 0? This in theory should\n",
    "            # never happen so we warn about it but deal\n",
    "            if k == reference_key and len(idr_dict[d][k]) == 0:\n",
    "                print(f'WARNING: Somehow  S. cervisiae IDR was length 0 for {k}. Skipping...')\n",
    "                tmp = {}\n",
    "                tmp_aligned\n",
    "                break\n",
    "\n",
    "            # if IDR has at least one residue\n",
    "            if len(idr_dict[d][k]) > 0:                \n",
    "                tmp[k] = idr_dict[d][k]\n",
    "                tmp_aligned[k] = idr_dict_aligned[d][k]\n",
    "\n",
    "        # predict disorder (note we use same metapredict version as used earlier!)\n",
    "        b = meta.predict_disorder(tmp, return_domains=True,silence_warnings=True, version=metapredict_version)\n",
    "\n",
    "        # finally filyer based on IDR length and fraction disordered\n",
    "        tmp2 = {}\n",
    "        for k in b:\n",
    "\n",
    "            # calculate what fraction of the IDR falls within a predicted disordered domain\n",
    "            fraction_disordered = np.sum([len(x) for x in b[k].disordered_domains])/len(b[k].sequence)\n",
    "            \n",
    "            # if we're on the S. cerevisiae protein include regardless of anything else...\n",
    "            if k == reference_key:\n",
    "                tmp2[k] = tmp_aligned[k]\n",
    "\n",
    "            \n",
    "            # else check the protein we've indentified is sufficiently disordered and sufficiently long\n",
    "            elif fraction_disordered > fraction_disordered_threshold and len(b[k].sequence) > min_idr_length:\n",
    "                tmp2[k] = tmp_aligned[k]\n",
    "        \n",
    "        idr_dict_filtered[d] = tmp2\n",
    "        \n",
    "    \n",
    "    return idr_dict_filtered\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56a1f48e-c077-4566-841c-d3a66b4a5c00",
   "metadata": {},
   "outputs": [],
   "source": [
    "ysn2idrvar = {}\n",
    "ysn2idrs = {}\n",
    "\n",
    "# for each YSN calculate variance in epsilon values across orthologous IDRs\n",
    "for ysn in tqdm(all_valid_IDs):\n",
    "\n",
    "    # get filtered orthologous IDRs associated with this YSN\n",
    "    orth_idrs = get_orthologous_idrs(ysn, fungi_matrix, ysn2disorder)\n",
    "    \n",
    "    idr2eps_var = {}\n",
    "    idr2eps_count = {}\n",
    "    tmp_ysn2idrs = {}\n",
    "\n",
    "    # for each orthologous IDR calculate variance in epsilon values across the orthologs\n",
    "    for idr_key in orth_idrs:    \n",
    "        ortholog_matrix = []\n",
    "        if len(orth_idrs[idr_key]) > 15:\n",
    "            tmp_ysn2idrs[idr_key] = orth_idrs[idr_key]\n",
    "            for k in orth_idrs[idr_key]:\n",
    "            \n",
    "                local_seq = orth_idrs[idr_key][k].replace('-','')\n",
    "                tmp = []\n",
    "                \n",
    "                for f in fingerprints.mpipi_fingerprints:    \n",
    "                    f_seq = fingerprints.mpipi_fingerprints[f]\n",
    "                    tmp.append(mf.epsilon(local_seq, f_seq)/len(local_seq))\n",
    "                ortholog_matrix.append(tmp)\n",
    "        \n",
    "            idr2eps_var[idr_key] = np.var(np.array(ortholog_matrix).transpose(),axis=1)    \n",
    "    \n",
    "    ysn2idrvar[ysn] = idr2eps_var\n",
    "    ysn2idrs[ysn] = tmp_ysn2idrs\n",
    "    \n",
    "    \n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96b9acd4-3647-499c-91d6-10da532112fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'generated_data/ysn2idrvar_metapredict_{metapredict_version}_disorder_{fraction_disordered}_{min_idr_length}.pkl', 'wb') as file:    \n",
    "    pickle.dump(ysn2idrvar, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddb5c9a2-bde8-4237-9106-11b5f311af79",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'generated_data/ysn2idrs_{metapredict_version}_{metapredict_version}_disorder_{fraction_disordered}_{min_idr_length}.pkl', 'wb') as file:    \n",
    "    pickle.dump(ysn2idrs, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "564a6a8e-3d71-4ae5-8e09-639b6f8e1117",
   "metadata": {},
   "source": [
    "## Finally generate sequence data\n",
    "Finally we save out FASTA files for the aligned and unaligned IDRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "157ecfa5-0fd6-4b52-9337-37f469a1aacd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NB set REBUILT to false cos we don't need to rebuild all these!\n",
    "REBUILD = False\n",
    "\n",
    "if REBUILD:\n",
    "\n",
    "    IN_PICKLE = f'generated_data/ysn2idrs_{metapredict_version}_{metapredict_version}_disorder_{fraction_disordered}_{min_idr_length}.pkl'\n",
    "    OUTDIR = f'idr_msas/metapredict_{metapredict_version}/full_protein_aligned/'\n",
    "    \n",
    "    # read in the generate IDR file\n",
    "    with open(IN_PICKLE, 'rb') as file:\n",
    "        data = pickle.load(file)\n",
    "\n",
    "\n",
    "    ## FIRST we write out the ALIGNED sequences using the \"full\" alignment (i.e. alignment\n",
    "    ## that was done on the full-length proteins\n",
    "    \n",
    "    for ysn in data:\n",
    "        for idr in data[ysn]:\n",
    "            protfasta.write_fasta(data[ysn][idr], f'{OUTDIR}/ALIGNED_{ysn}_{idr}.fasta')\n",
    "    \n",
    "    ## THEN we write out JUST the IDRs from the othologs - these sequences can then be \n",
    "    ## re-aligned post-factor using clustal-omega\n",
    "    OUTDIR = f'idr_msas/metapredict_{metapredict_version}/'\n",
    "    for ysn in data:\n",
    "        for idr in data[ysn]:\n",
    "            tmp = {}\n",
    "            for k in data[ysn][idr]:\n",
    "                tmp[k] = data[ysn][idr][k].replace('-','')\n",
    "            protfasta.write_fasta(tmp, f'{OUTDIR}/{ysn}_{idr}.fasta')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
