# About

## Files
NOTE that on GitHub, the directories described here are excluded because they are too large. However, these are provided in the Zenodo repository. 

## Description
This directory contains subdirectories for IDRs identified using metapredict (v1 and v3).

Practically, these files are generated by `../2_build_orthologous_IDRs.ipynb` and done by:


1. Finding orthlogs using the YGOB data (as done in Figure 1).
2. Identifying all IDRs in S. cerevisiae 
3. Use those boundaries to find equivalent regions in all the other orthologs and ensure those newly-found IDRs are at least 80% disordered and over 50 residues in length.
4. Saving out those orthologous IDRs in their full-length aligned format (one .fasta file per IDR, saved in `metapredict_<version>/full_protein_aligned/`) or unaligned format, saved in `metapredict_<version>/`.
5. FINALLY, we can take the unaligned IDRs in `metapredict_<version>/` and align each IDR within that file using Clustal Omega on a per-IDR basis; these are saved in `metapredict_<version>/IDR_aligned/`.

Files have the format `<YSN>_IDR<IDR_ID>_<START>_<END>.fasta` (e.g. `YFL007W_IDR2_181_235.fasta`). 

Each of these has at least 16 or more sequences (up to a maximum of 32). For Clustal-based alignment of individual IDRs, we used Clustal with default settings, e.g.;

	clustal-omega-1.2.3-macosx -i YFL007W_IDR2_181_235.fasta > YFL007W_IDR2_181_235