{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2cb5a065-f791-4815-88b5-9e3c1858d9d4",
   "metadata": {},
   "source": [
    "## Synthetic IDR evolution\n",
    "NOTE - this notebook probably doesn't actually need to be run..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4af5cfa2-2210-4fa3-973b-a6519a8da205",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pyplot import figure\n",
    "\t\n",
    "# Set such that PDF fonts export in a manner that they\n",
    "# are editable in illustrator/affinity\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "# set to define axes linewidths\n",
    "matplotlib.rcParams['axes.linewidth'] = 0.5\n",
    "\n",
    "# this defines some prefactors so inline figures look nice\n",
    "# on a retina macbook. These can be commented out without any\n",
    "# issue and are solely asthetic.\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format='retina'\n",
    "\n",
    "# UPDATE 2020-12-31 (my preferred font is Avenir...)\n",
    "font = {'family' : 'avenir',\n",
    "    \t'weight' : 'normal'}\n",
    "\n",
    "matplotlib.rc('font', **font)\n",
    "from sparrow import Protein\n",
    "from sparrow.predictors import batch_predict\n",
    "\n",
    "import protfasta\n",
    "\n",
    "from finches import CALVADOS_frontend, Mpipi_frontend\n",
    "\n",
    "mf = Mpipi_frontend()\n",
    "cf = CALVADOS_frontend()\n",
    "from sparrow import Protein\n",
    "\n",
    "from shephard import interfaces, apis\n",
    "from housetools import nucleic_acid_tools \n",
    "\n",
    "from mutagenesis import run_mutagenesis\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "def compare(s1,s2, verbose=True):\n",
    "    count_diffs = 0\n",
    "    for i in range(len(s1)):\n",
    "        if s1[i] != s2[i]:\n",
    "            count_diffs = count_diffs + 1\n",
    "            if verbose:\n",
    "                print(f\"{i+1}: {s1[i]} vs. {s2[i]}\")\n",
    "\n",
    "    if verbose:\n",
    "        print(f\"{count_diffs} changes\")\n",
    "    return count_diffs\n",
    "\n",
    "def average_mutations(lib):\n",
    "    vals = []\n",
    "    for k in lib:\n",
    "\n",
    "        # nb abf1_idr must be defined outside!\n",
    "        vals.append(compare(abf1_idr, lib[k], verbose=False))\n",
    "    return np.mean(vals)\n",
    "\n",
    "idr2 = 'NNNNNNDGELSGTNLRSNSIDYAKHQEISSAGTSSNTTKNVNNNKNDSNDDNNGNNNNDASNLMESVLDKTSSHRYQPKKMPSVNKWSKPDQITHSDVSMVGLDESNDGGNENVHPTLAEVDAQEARETAQLAIDKINSYKRSIDDKNGDGHNNSSRNVVDENLINDMDSEDAHKSKRQHLSDITLEERNEDDKLPHEVAEQLRLLSSHLKEVE'\n",
    "\n",
    "from housetools.sequence_tools.sequence_conservation import ConservationCalculator\n",
    "\n",
    "from shephard.apis import uniprot\n",
    "yp = uniprot.uniprot_fasta_to_proteome('data/s_cerevisiae_clean.fasta')\n",
    "interfaces.si_domains.add_domains_from_file(yp,'data/shprd_domains_idrs_s_cerevisiae.tsv')\n",
    "interfaces.si_protein_attributes.add_protein_attributes_from_file(yp, 'data/abf1_interactors_stringdb.tsv')\n",
    "\n",
    "mf = Mpipi_frontend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d86ab1b-2c37-414c-b5f0-a864a10c7d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "abf1 = yp.protein('P14164')\n",
    "\n",
    "rad7_idr = 'MYRSRNRPKRGGENEVKGPNSALTQFLREEGISAENIKQKWYQRQSKKQEDATDEKKGKAEDDSFTAEISRVVEDEEIDEIGTGSGTETERAQVSYDARMKLVPADSDEEEYETSHISDTPVSLSSANDRESLTKKRQNTAKIIQNRRRKRKRAADLLDRRVNKVSSLQSLCITKISENISKWQKEADESSKLVFNKLRDVLGGVST'\n",
    "abf1_idr = 'NNNNNNDGELSGTNLRSNSIDYAKHQEISSAGTSSNTTKNVNNNKNDSNDDNNGNNNNDASNLMESVLDKTSSHRYQPKKMPSVNKWSKPDQITHSDVSMVGLDESNDGGNENVHPTLAEVDAQEARETAQLAIDKINSYKRSIDDKNGDGHNNSSRNVVDENLINDMDSEDAHKSKRQHLSDITLEERNEDDKLPHEVAEQLRLLSSHLKEVE'\n",
    "\n",
    "x = mf.intermolecular_idr_matrix(abf1_idr, rad7_idr)\n",
    "\n",
    "abf1_idr1_attractive_interaction_score = np.sum(x[0][0][x[0][0] < 0])\n",
    "print(abf1_idr1_attractive_interaction_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3002bd84-99da-45a4-b2b1-83608dfd725e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the code here selects some possible additional\n",
    "\n",
    "hits = []\n",
    "\n",
    "# this is a list of uniprot IDs for the STRINGDM-defined interactors for Abf1\n",
    "abf1_interactors = abf1.attribute('partners').split(', ')\n",
    "\n",
    "# for each interactor\n",
    "for interactor in abf1_interactors:\n",
    "\n",
    "    # for each IDR in the interactor \n",
    "    for idr in yp.protein(interactor).domains:     \n",
    "\n",
    "        # make sure IDRs are long enough!!\n",
    "        if len(idr.sequence) > 80:\n",
    "\n",
    "            # calculate the interaction matrix for IDR2 to the normal sequence\n",
    "            x = mf.intermolecular_idr_matrix(abf1_idr, idr.sequence)    \n",
    "\n",
    "            # safe=False means we can overwrite if we re-run this - collapse down all the\n",
    "            # attractive interactions only\n",
    "            idr.add_attribute('idr2_interaction_score', np.sum(x[0][0][x[0][0] < 0]), safe=False)\n",
    "\n",
    "            # if IDRs are between 80 and 170 in length...\n",
    "            if len(idr.sequence) > 80 and len(idr.sequence) < 170:\n",
    "\n",
    "                # and if IDRs are attractive enough (-2000 is a relatively arbitrary define)\n",
    "                if idr.attribute('idr2_interaction_score') < -2000:\n",
    "                    hits.append(idr)\n",
    "\n",
    "# generate intermaps for those hits\n",
    "for idr in hits:\n",
    "    mf.interaction_figure(abf1_idr, idr.sequence, fname=f'figures/{idr.protein.unique_ID}_{idr.start}_{idr.end}.pdf', zero_folded=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33a3fa08-2a6a-47c4-9343-2a62f92f94c3",
   "metadata": {},
   "source": [
    "## Selection without compensation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a9b9a8f-050d-4705-8a8f-7cdf9e28557b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in hits:\n",
    "    print(i)\n",
    "    print(i.sequence)\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3b742f8-dd34-4ec7-9b27-b350f9097d50",
   "metadata": {},
   "source": [
    "### Build synthetic libraries\n",
    "Cells below generate the synthetic libraries of orthologs either (real library) under selection or with no selection (null_library, uses a BIG delta threshold so everything gets accepted)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1edc52fd-c6cb-4e6f-a89d-59c6afe41d38",
   "metadata": {},
   "outputs": [],
   "source": [
    "libsize = 100\n",
    "\n",
    "for idr in tqdm(hits[1:]):\n",
    "\n",
    "    # null library\n",
    "    null_library = run_mutagenesis(idr2, \n",
    "                                   idr.sequence,\n",
    "                                   num_mutants=libsize, \n",
    "                                   delta_threshold=2000, \n",
    "                                   num_loops=14,\n",
    "                                   target_size=30,\n",
    "                                   number_of_events=3,\n",
    "                                   mode='mpipi',\n",
    "                                   verbose=True)\n",
    "\n",
    "\n",
    "    real_library = run_mutagenesis(idr2, \n",
    "                                   idr.sequence,\n",
    "                                   num_mutants=libsize,\n",
    "                                   delta_threshold=50, \n",
    "                                   num_loops=19,\n",
    "                                   target_size=30,\n",
    "                                   number_of_events=3,\n",
    "                                   mode='mpipi',\n",
    "                                   selection_type='attractive_sumdiff',                                    \n",
    "                                   verbose=True)\n",
    "\n",
    "\n",
    "    null_average = round(average_mutations(null_library))\n",
    "    real_average = round(average_mutations(real_library))\n",
    "\n",
    "    protfasta.write_fasta(null_library, f'out_libraries/null_{idr.protein.unique_ID}_{idr.start}_{idr.end}_size_{len(null_library)}_muts_{null_average}.fasta')\n",
    "    protfasta.write_fasta(real_library, f'out_libraries/real_{idr.protein.unique_ID}_{idr.start}_{idr.end}_size_{len(real_library)}_muts_{real_average}.fasta')\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
