{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "303b2153",
   "metadata": {},
   "source": [
    "# About\n",
    "This notebook generates the synthetic IDR libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4af5cfa2-2210-4fa3-973b-a6519a8da205",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/alex/.uv/neuron_uv/.venv/lib/python3.12/site-packages/finches/forcefields/calvados.py:236: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.038286503882254706' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  r.loc['H','q'] = 1. / ( 1 + 10**(self.pH-6) )\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pyplot import figure\n",
    "\t\n",
    "# Set such that PDF fonts export in a manner that they\n",
    "# are editable in illustrator/affinity\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "# set to define axes linewidths\n",
    "matplotlib.rcParams['axes.linewidth'] = 0.5\n",
    "\n",
    "# this defines some prefactors so inline figures look nice\n",
    "# on a retina macbook. These can be commented out without any\n",
    "# issue and are solely asthetic.\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format='retina'\n",
    "\n",
    "# UPDATE 2020-12-31 (my preferred font is Avenir...)\n",
    "font = {'family' : 'avenir',\n",
    "    \t'weight' : 'normal'}\n",
    "\n",
    "matplotlib.rc('font', **font)\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "import protfasta\n",
    "from finches import CALVADOS_frontend, Mpipi_frontend\n",
    "\n",
    "mf = Mpipi_frontend()\n",
    "cf = CALVADOS_frontend()\n",
    "\n",
    "\n",
    "# NB: this function requires the housetools package, which is a holehouselab\n",
    "# specific package. If you want to run this code yourself, please contact us\n",
    "# so we can provide you with a newer version of this code!\n",
    "from mutagenesis import run_mutagenesis, compare\n",
    "\n",
    "\n",
    "# define the IDRs we are going to use for the figure 5 library generation\n",
    "idr2 = 'NNNNNNDGELSGTNLRSNSIDYAKHQEISSAGTSSNTTKNVNNNKNDSNDDNNGNNNNDASNLMESVLDKTSSHRYQPKKMPSVNKWSKPDQITHSDVSMVGLDESNDGGNENVHPTLAEVDAQEARETAQLAIDKINSYKRSIDDKNGDGHNNSSRNVVDENLINDMDSEDAHKSKRQHLSDITLEERNEDDKLPHEVAEQLRLLSSHLKEVE'\n",
    "rad7_idr = 'MYRSRNRPKRGGENEVKGPNSALTQFLREEGISAENIKQKWYQRQSKKQEDATDEKKGKAEDDSFTAEISRVVEDEEIDEIGTGSGTETERAQVSYDARMKLVPADSDEEEYETSHISDTPVSLSSANDRESLTKKRQNTAKIIQNRRRKRKRAADLLDRRVNKVSSLQSLCITKISENISKWQKEADESSKLVFNKLRDVLGGVST'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33a3fa08-2a6a-47c4-9343-2a62f92f94c3",
   "metadata": {},
   "source": [
    "## Selection without compensation\n",
    "For the libraries built here, the selection type used is the default (`maxdiff`), which means that the allowable maximum data is the difference between any pair of elements in the interaction matrix. This means no pair of residues can lead to a big change and no compensatory changes are tolerated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1edc52fd-c6cb-4e6f-a89d-59c6afe41d38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in existing sequence libraries...\n"
     ]
    }
   ],
   "source": [
    "rebuild_library = False\n",
    "\n",
    "# if True, will rebuild the sequence libraries\n",
    "if rebuild_library:\n",
    "\n",
    "    print(\"Rebuilding sequence libraries...\")\n",
    "\n",
    "    # run mutagenesis with specified parameters. These settings\n",
    "    # should yield ~28 mutations on average\n",
    "    new_seq_library = run_mutagenesis(idr2, \n",
    "                    rad7_idr,\n",
    "                    num_mutants=350, \n",
    "                    delta_threshold=0.5, \n",
    "                    num_loops=20,\n",
    "                    target_size=30,\n",
    "                    number_of_events=3,\n",
    "                    mode='mpipi',\n",
    "                    verbose=True)\n",
    "\n",
    "    # should be 28 mutations on average\n",
    "    protfasta.write_fasta(new_seq_library, f'chemlib_{len(new_seq_library)}.fasta')\n",
    "\n",
    "    new_seq_library_tight = run_mutagenesis(idr2, \n",
    "                                      rad7_idr,\n",
    "                                      num_mutants=350, \n",
    "                                      delta_threshold=0.1, \n",
    "                                      num_loops=30,\n",
    "                                      target_size=30,\n",
    "                                      number_of_events=3,\n",
    "                                      mode='mpipi',\n",
    "                                      verbose=True)\n",
    "\n",
    "    # should be 22 mutations on average\n",
    "    protfasta.write_fasta(new_seq_library_tight, f'chemlib_{len(new_seq_library_tight)}_tight.fasta')\n",
    "\n",
    "\n",
    "    new_seq_library_null_28 = run_mutagenesis(idr2, \n",
    "                                      rad7_idr,\n",
    "                                      num_mutants=350, \n",
    "                                      delta_threshold=20, \n",
    "                                      num_loops=14,\n",
    "                                      target_size=30,\n",
    "                                      number_of_events=3,\n",
    "                                      mode='mpipi',\n",
    "                                      verbose=True)\n",
    "    \n",
    "    # should be 28 on average\n",
    "    protfasta.write_fasta(new_seq_library_null_28, f'chemlib_{len(new_seq_library_tight)}_null_28.fasta')\n",
    "\n",
    "\n",
    "    new_seq_library_null_21 = run_mutagenesis(idr2, \n",
    "                                      rad7_idr,\n",
    "                                      num_mutants=350, \n",
    "                                      delta_threshold=20, \n",
    "                                      num_loops=11,\n",
    "                                      target_size=30,\n",
    "                                      number_of_events=3,\n",
    "                                      mode='mpipi',\n",
    "                                      verbose=True)\n",
    "    \n",
    "    # should be 21 on average\n",
    "    protfasta.write_fasta(new_seq_library_null_21, f'chemlib_{len(new_seq_library_tight)}_null_21.fasta')\n",
    "\n",
    "\n",
    "\n",
    "else:\n",
    "    print(\"Reading in existing sequence libraries...\")\n",
    "    new_seq_library = protfasta.read_fasta('libraries/chemlib_350.fasta')\n",
    "    new_seq_library_tight = protfasta.read_fasta('libraries/chemlib_350_tight.fasta')\n",
    "\n",
    "    # note - THIS is going to be the null we use (has 28 mutations on average), but note\n",
    "    # that even though we're gonna compare this with the selection done using attractive_sumdiff\n",
    "    # insteadf of maxdiff, because we're applying NO selection it doesn't matter the selection\n",
    "    # regime we're using as everything gets accepted.\n",
    "    new_seq_library_null_28 = protfasta.read_fasta('libraries/chemlib_350_null_28.fasta')\n",
    "    new_seq_library_null_21 = protfasta.read_fasta('libraries/chemlib_350_null_21.fasta')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e7ceb6c-5b8f-4ca8-a826-1f1677debf49",
   "metadata": {},
   "source": [
    "## Selection with compensation\n",
    "Here we define the `selection_type` as mode `attractive_sumdiff`, which means the different allowed reflects the sum of the differences between the interaction matrix of the wild type and the mutant, but only for the attractive interactions (that is, increase in repuslive interactions is tolerated as long as not at expense of attractive interactions, and attraction can be redistributed across the sequence). Note that the numerical `delta_threshold` value must change depending on the `selection_type` (i.e. here it's 50, which is MUCH less permissive than 20 if `maxdiff` were used, for example)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdad5a3a-2fb7-4990-9969-7f496e9b0631",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in existing sequence libraries...\n"
     ]
    }
   ],
   "source": [
    "rebuild_library = False\n",
    "\n",
    "if rebuild_library:\n",
    "\n",
    "    # this gives ~28 mutations per sequence\n",
    "    new_seq_library_v2 = run_mutagenesis(idr2, \n",
    "                    rad7_idr,\n",
    "                    num_mutants=350, \n",
    "                    delta_threshold=50, \n",
    "                    num_loops=19,\n",
    "                    target_size=30,\n",
    "                    number_of_events=3,\n",
    "                    mode='mpipi',\n",
    "                    selection_type='attractive_sumdiff',                                    \n",
    "                    verbose=True)\n",
    "\n",
    "    protfasta.write_fasta(new_seq_library_v2, f'chemlib_v2_{len(new_seq_library_v2)}.fasta')\n",
    "\n",
    "else:\n",
    "    print(\"Reading in existing sequence libraries...\")\n",
    "\n",
    "    \n",
    "    new_seq_library_v2 = protfasta.read_fasta('libraries/chemlib_v2_350.fasta')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b70367be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the library of interest for the remainder of the analysis\n",
    "LIBRARY_OF_INTEREST = new_seq_library_v2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "064b554a",
   "metadata": {},
   "source": [
    "### Write intermaps out\n",
    "If we set `WRITE_MAPS` to true, the cell below will generate an intermap image for each synthetic ortholog to enable easy visualization of how these variants change the instantaneous intermap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a4d01744-c5e6-41ac-8465-02ab2440c29d",
   "metadata": {},
   "outputs": [],
   "source": [
    "WRITE_MAPS = False\n",
    "\n",
    "if WRITE_MAPS:\n",
    "\n",
    "    # this will take 2-3 minutes to run \n",
    "    for k in tqdm(LIBRARY_OF_INTEREST)  :\n",
    "        s = LIBRARY_OF_INTEREST[k]\n",
    "        mf.interaction_figure(s, rad7_idr, no_disorder=True,zero_folded=False)\n",
    "        plt.savefig(f'figures/individual_intermaps/idr2_rad7_intermap_{k}.png')\n",
    "        plt.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "496ce471",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Real library: 27.69 +/- 4.13 mutations per sequence\n",
      "Null library: 27.83 +/- 2.91 mutations per sequence\n"
     ]
    }
   ],
   "source": [
    "mutations_real = []\n",
    "mutations_null = []\n",
    "for k in new_seq_library_v2:\n",
    "    mutations_real.append(compare(idr2, new_seq_library_v2[k], verbose=False))\n",
    "\n",
    "for k in new_seq_library_null_28:    \n",
    "    mutations_null.append(compare(idr2, new_seq_library_null_28[k], verbose=False))\n",
    "    \n",
    "print(f\"Real library: {np.mean(mutations_real):.2f} +/- {np.std(mutations_real):.2f} mutations per sequence\")    \n",
    "print(f\"Null library: {np.mean(mutations_null):.2f} +/- {np.std(mutations_null):.2f} mutations per sequence\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f8b8abb",
   "metadata": {},
   "source": [
    "## Large library construction\n",
    "Finally, we go ahead and generate a large library. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7fc1aaad-d156-4148-9537-13302c4c749d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in existing sequence libraries...\n"
     ]
    }
   ],
   "source": [
    "# BIG LIBRARIES\n",
    "rebuild_library = False\n",
    "\n",
    "if rebuild_library:\n",
    "    new_seq_library_big = run_mutagenesis(idr2, \n",
    "                    rad7_idr,\n",
    "                    num_mutants=650, \n",
    "                    delta_threshold=0.5, \n",
    "                    num_loops=20,\n",
    "                    target_size=30,\n",
    "                    number_of_events=3,\n",
    "                    mode='mpipi',\n",
    "                    verbose=True)\n",
    "\n",
    "    # should be 28 mutations on average\n",
    "    protfasta.write_fasta(new_seq_library_big, f'chemlib_{len(new_seq_library_big)}.fasta')\n",
    "\n",
    "\n",
    "    new_seq_library_null_28_big = run_mutagenesis(idr2, \n",
    "                                      rad7_idr,\n",
    "                                      num_mutants=650, \n",
    "                                      delta_threshold=20, \n",
    "                                      num_loops=14,\n",
    "                                      target_size=30,\n",
    "                                      number_of_events=3,\n",
    "                                      mode='mpipi',\n",
    "                                      verbose=True)\n",
    "    \n",
    "    # should be 28 on average\n",
    "    protfasta.write_fasta(new_seq_library_null_28_big, f'chemlib_{len(new_seq_library_tight)}_null_28.fasta')\n",
    "else:\n",
    "    print(\"Reading in existing sequence libraries...\")\n",
    "    new_seq_library_big = protfasta.read_fasta('libraries/chemlib_650.fasta')\n",
    "    new_seq_library_null_28_big = protfasta.read_fasta('libraries/chemlib_650_null_28.fasta')\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "neuron_uv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
