{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3e20440e",
   "metadata": {},
   "source": [
    "### Supplementary Figure 9\n",
    "###### last updated 2024-10-28\n",
    "This notebook analyzes all human TFs and generates panels for supplementary figure 9.\n",
    "\n",
    "This notebook requires [metapredict](https://metapredict.readthedocs.io/en/latest/) and [shephard](https://shephard.readthedocs.io/en/latest/). Both packages were developed by the Holehouse lab and can be installed using `pip` via\n",
    "\n",
    "    pip install metapredict, shephard\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f84ec41",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pyplot import figure\n",
    "\t\n",
    "# Set such that PDF fonts export in a manner that they\n",
    "# are editable in illustrator/affinity\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "# set to define axes linewidths\n",
    "matplotlib.rcParams['axes.linewidth'] = 0.5\n",
    "\n",
    "# this defines some prefactors so inline figures look nice\n",
    "# on a retina macbook. These can be commented out without any\n",
    "# issue and are solely asthetic.\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format='retina'\n",
    "\n",
    "# UPDATE 2020-12-31 (my preferred font is Avenir...)\n",
    "font = {'family' : 'arial',\n",
    "    \t'weight' : 'normal'}\n",
    "\n",
    "matplotlib.rc('font', **font)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6826f8cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from shephard import interfaces \n",
    "from shephard.apis import uniprot\n",
    "from shephard.apis import metapredict_api\n",
    "import metapredict as meta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "373d460b-86d5-40cf-ad47-22cfb2cf36c3",
   "metadata": {},
   "source": [
    "### Annotate with IDRs across all TFs and across human proteome\n",
    "NOTE this requires the most recent version of SHEPHARD that will only be released once metapredict V3 is pushed to master and on PyPI."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c6f4e76-e11e-44c9-b572-7ab105c5ab54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# nb - if you want to recompute the IDRs you can, but we have precomputed everything for you...\n",
    "RECOMPUTE = False\n",
    "SAVE_DATA = False\n",
    "\n",
    "\n",
    "# Read in the set of human transcription factors as a SHEPHARD Proteome object\n",
    "TFs = uniprot.uniprot_fasta_to_proteome('human_TFs_full_2024_10_29.fasta')\n",
    "\n",
    "# note we have the invalid_sequence_action='convert' as there are a couple of 'X' residues in\n",
    "# human proteome we convert.\n",
    "HP  = uniprot.uniprot_fasta_to_proteome('UP000005640_9606.fasta', invalid_sequence_action='convert')\n",
    "\n",
    "\n",
    "# if recompute we recalculate all IDRs\n",
    "if RECOMPUTE:\n",
    "    # ~4 seconds\n",
    "    metapredict_api.annotate_proteome_with_disordered_domains(TFs)\n",
    "\n",
    "    # ~60 seconds\n",
    "    metapredict_api.annotate_proteome_with_disordered_domains(HP)\n",
    "\n",
    "    # if save data save these annotations\n",
    "    if SAVE_DATA:\n",
    "        interfaces.si_domains.write_domains(TFs, 'data/TF_IDRs_v3.tsv')\n",
    "        interfaces.si_domains.write_domains(HP, 'data/HP_IDRs_v3.tsv')\n",
    "\n",
    "        # save human TF UIDs\n",
    "        fh = open('human_tf_uniprot_IDs.txt','w')\n",
    "        for protein in TFs.proteins:\n",
    "            fh.write(f\"{protein}\\n\")\n",
    "        fh.close()    \n",
    "\n",
    "\n",
    "\n",
    "else:\n",
    "    interfaces.si_domains.add_domains_from_file(TFs, 'data/TF_IDRs_v3.tsv')\n",
    "    interfaces.si_domains.add_domains_from_file(HP, 'data/HP_IDRs_v3.tsv')\n",
    "\n",
    "print('Finished loading...')\n",
    "print(f\"Human proteome loaded; {len(HP)} proteins and {len(HP.domains)} IDRs\")\n",
    "print(f\"Human transcription factors loaded; {len(TFs)} proteins and {len(TFs.domains)} IDRs\")    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83882776",
   "metadata": {},
   "source": [
    "### Compute fraction of disorder \n",
    "The cell below cycles through each transcription factor and calculates what fraction of each protein is disordered or folded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8be46e5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fraction_disordered = []\n",
    "\n",
    "number_disordered = []\n",
    "number_ordered = []\n",
    "prot_len = []\n",
    "\n",
    "for protein in TFs:\n",
    "    \n",
    "    \n",
    "    # generates a list with the length of each IDR in the protein\n",
    "    IDR_lengths = [len(d) for d in protein.domains if d.domain_type == 'IDR' ]\n",
    "    \n",
    "    # calculates the fraction of the protein in the IDR\n",
    "    fraction_disordered.append(np.sum(IDR_lengths)/len(protein))\n",
    "    number_disordered.append(np.sum(IDR_lengths))\n",
    "    number_ordered.append(len(protein)-np.sum(IDR_lengths))\n",
    "    prot_len.append(len(protein))\n",
    "    \n",
    "\n",
    "sorted_lists = sorted(zip(prot_len, number_disordered, number_ordered),reverse=True)\n",
    "\n",
    "# Unzip the sorted lists\n",
    "sorted_list1, sorted_list2, sorted_list3 = zip(*sorted_lists)\n",
    "\n",
    "# Convert back to lists if needed\n",
    "prot_len = list(sorted_list1)\n",
    "number_disordered = list(sorted_list2)\n",
    "number_ordered = list(sorted_list3)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cb6fdf9-424e-4ecf-b671-81f7676c7d3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the plot\n",
    "fig, ax = plt.subplots(figsize=(8,5))\n",
    "x = np.arange(len(sorted_list2))\n",
    "\n",
    "ax.barh(x, number_disordered, label='disorder', color='r', height=1)\n",
    "ax.barh(x, number_ordered, left=number_disordered, label='folded', color='k', height=1)  # Stacked on top\n",
    "\n",
    "# Set labels and title\n",
    "ax.set_xlabel('Protein length')\n",
    "ax.set_ylabel('Protein index')\n",
    "plt.ylim([x[0],x[-1]])\n",
    "\n",
    "\n",
    "ax.legend(frameon=False)\n",
    "plt.savefig('figures/human_TFs.pdf')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91ccb779",
   "metadata": {},
   "source": [
    "### Violin plot of disorder\n",
    "The cell below generates and saves a violin plot of fractional disorder for human proteome vs. TFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4411057e-121c-400e-aae1-a8d8b00e5002",
   "metadata": {},
   "outputs": [],
   "source": [
    "fraction_disordered_HP = []\n",
    "for protein in HP:\n",
    "    \n",
    "    \n",
    "    # generates a list with the length of each IDR in the protein\n",
    "    IDR_lengths = [len(d) for d in protein.domains if d.domain_type == 'IDR' ]\n",
    "    \n",
    "    # calculates the fraction of the protein in the IDR\n",
    "    fraction_disordered_HP.append(np.sum(IDR_lengths)/len(protein))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dca3be32",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure(num=None, figsize=(4, 3), dpi=300, facecolor='w', edgecolor='k')\n",
    "ax = plt.gca()\n",
    "\n",
    "\n",
    "count = len(fraction_disordered_HP)\n",
    "plt.plot(np.array([1]*count)+(np.random.random(count)-0.5)*0.3, fraction_disordered_HP,'.',  color='red', alpha=0.1, ms=2, mew=0)\n",
    "\n",
    "count = len(fraction_disordered)\n",
    "plt.plot(np.array([2]*count)+(np.random.random(count)-0.5)*0.3, fraction_disordered, '.', color='orange', alpha=0.8, ms=2, mew=0)\n",
    "\n",
    "plt.xlim([0,3])\n",
    "parts = ax.violinplot([fraction_disordered_HP, fraction_disordered], showmeans=True)\n",
    "\n",
    "bodies_new = parts['bodies'][0]\n",
    "bodies_new.set_color('red')\n",
    "bodies_new.set_linewidth(0)\n",
    "\n",
    "bodies_new = parts['bodies'][1]\n",
    "bodies_new.set_color('orange')\n",
    "bodies_new.set_linewidth(0)\n",
    "\n",
    "\n",
    "for partname in ('cbars','cmins','cmaxes','cmeans'):\n",
    "\n",
    "    parts[partname].set_color('k')\n",
    "    parts[partname].set_edgecolor('k')\n",
    "    parts[partname].set_linewidth(0.5)\n",
    "\n",
    "plt.xlim([0.5,2.5])    \n",
    "plt.ylim([0,1.1])\n",
    "plt.text(0.82,1.05,f'median = {round(np.median(fraction_disordered_HP),3)}',fontsize=6)\n",
    "plt.text(1.82,1.05,f'median = {round(np.median(fraction_disordered),3)}',fontsize=6)\n",
    "\n",
    "plt.xticks([1,2],['Human\\nproteome','Transcription\\nfactors'], fontsize=8)\n",
    "plt.yticks(fontsize=7)\n",
    "plt.ylabel('Fraction disordered', fontsize=8)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('figures/proteome_vs_tfs.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63554f48-e04d-4534-baeb-8dbe291a9f26",
   "metadata": {},
   "outputs": [],
   "source": [
    "AAs = ['A','L','V','I','M','F','Y','W','P','S','G', 'Q','T','N','H','E','D','R','K','C']\n",
    "\n",
    "proteome_idrs = {}\n",
    "tf_idrs = {}\n",
    "\n",
    "proteome = {} \n",
    "tfs = {}\n",
    "\n",
    "for aa in AAs:\n",
    "    proteome_idrs[aa] = 0\n",
    "    tf_idrs[aa] = 0\n",
    "    proteome[aa] = 0\n",
    "    tfs[aa] = 0\n",
    "\n",
    "for protein in HP:\n",
    "    for d in protein.domains:\n",
    "        for aa in d.sequence:\n",
    "            proteome_idrs[aa] = proteome_idrs[aa] + 1\n",
    "\n",
    "    for aa in protein.sequence:\n",
    "        proteome[aa] = proteome[aa] + 1\n",
    "\n",
    "# all TF stuff\n",
    "for protein in TFs:\n",
    "    for d in protein.domains:\n",
    "        for aa in d.sequence:\n",
    "            tf_idrs[aa] = tf_idrs[aa] + 1\n",
    "\n",
    "    for aa in protein.sequence:\n",
    "        tfs[aa] = tfs[aa] + 1\n",
    "\n",
    "\n",
    "proteome_idrs = {aa:proteome_idrs[aa]/np.sum(list(proteome_idrs.values())) for aa in proteome_idrs}\n",
    "proteome = {aa:proteome[aa]/np.sum(list(proteome.values())) for aa in proteome}\n",
    "\n",
    "tf_idrs = {aa:tf_idrs[aa]/np.sum(list(tf_idrs.values())) for aa in tf_idrs}\n",
    "tfs = {aa:tfs[aa]/np.sum(list(tfs.values())) for aa in tfs}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9bb2ba0-53a2-4d1c-9d1e-c00fd5fcc2ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example data setup\n",
    "keys = list(proteome.keys())  # Assuming all dictionaries have the same keys\n",
    "proteome_values = [proteome[key] for key in keys]\n",
    "proteome_idrs_values = [proteome_idrs[key] for key in keys]\n",
    "tfs_values = [tfs[key] for key in keys]\n",
    "tf_idrs_values = [tf_idrs[key] for key in keys]\n",
    "\n",
    "# Parameters for plotting\n",
    "num_keys = len(keys)\n",
    "x = np.arange(num_keys)  # X-axis positions for each key\n",
    "bar_width = 0.2  # Width of each bar within a group\n",
    "\n",
    "# Create the plot\n",
    "fig, ax = plt.subplots(figsize=(7, 4))\n",
    "\n",
    "# Plot each dictionary's values as separate bars within each group\n",
    "ax.bar(x - 1.5 * bar_width, proteome_values, width=bar_width, label=\"Human proteome\", color='k')\n",
    "ax.bar(x - 0.5 * bar_width, proteome_idrs_values, width=bar_width, label=\"Human IDRs\", color='r', alpha=0.5)\n",
    "ax.bar(x + 0.5 * bar_width, tfs_values, width=bar_width, label=\"Human TFs\", color='grey')\n",
    "ax.bar(x + 1.5 * bar_width, tf_idrs_values, width=bar_width, label=\"Human TF IDRs\", color='orange', alpha=0.8)\n",
    "\n",
    "# Labeling\n",
    "ax.set_xlabel(\"Amino acids\", fontsize=12)\n",
    "ax.set_ylabel(\"Fraction Values\")\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(keys, rotation=0, ha=\"center\", fontsize=9)\n",
    "ax.legend(frameon=False)\n",
    "\n",
    "# Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig('figures/composition_comparison.pdf')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
