# Tables for Lotthammer, Ginell, Griffith et al. 2023
#### Last updated 2023-12-25


All supplementary tables provided as .xlsx or .csv files for convenience. 

**Table S1.** Table showing proteins with overly-compact subregions ordered by protein abundance.
Protein abundance data is from quantitative proteomics mass spectrometry experiments previously published by Hein et al. and parsed and built into SHEPHARD-compliant protein attribute annotation files. See separate .xlsx file, or data individual file at the GitHub repository.

**Table S2.**** Table showing proteins with overly-expanded subregions ordered by protein abundance.
Protein abundance data is from quantitative proteomics mass spectrometry experiments previously published by Hein et al. and parsed and built into SHEPHARD-compliant protein attribute annotation files. See separate .xlsx file, or data individual file at the GitHub repository.

**Table S3**. Gene ontology analysis for proteins with compact IDRs. For annotations that contained 100 or more entries in the basis set and showed 2-fold or higher enrichment, the vast majority of annotated terms pertain to RNA in some way across the three classes of gene ontology. PANTHER reports using Fisher’s exact test (default behavior, two-sided). See separate .xlsx file, or data individual file at the GitHub repository.

**Table S4**. Gene ontology analysis for proteins with expanded IDRs. For annotations that contained 100 or more entries in the basis set and showed 2-fold or higher enrichment, a variety of IDR-associated annotations are identified, including chromatin binding, cytoskeletal regulation, and cellular organization, in good agreement with analogous analysis from Tesei & Trolle, despite the two analyses being done in different ways. PANTHER reports using Fisher’s exact test (default behavior, two-sided). See separate .xlsx file, or data individual file at the GitHub repository.

**Table S5**. Yak1 homologs. Yak1 homolog IDR sequences, along with organism, reference ID, residue positions, and ALBATROSS-predicted Re.

**Table S6**. Spt2 homologs. Spt2 homolog IDR sequences, organism, reference ID, residue positions, and ALBATROSS-predicted Re.

**Table S7**. Final Network Hyperparameters. The best-performing hyperparameters used in the final ALBATROSS implementation for each polymeric property predictor are shown below.

**Table S8**. Experimental and prediction data used in Supplementary Figure S8. See separate .xlsx file, or data individual file at the GitHub repository.


